crates/provider/src/shared/client.rs

```diff
@@ -0,0 +1,71 @@
+use std::error::Error;
+
+use async_trait::async_trait;
+
+use super::types::{FileMap, NameSpaceDef, PodDef, Settings};
+use crate::shared::types::{RunCommandOptions, RunCommandResponse};
+
+#[async_trait]
+#[allow(non_upper_case_globals)]
+pub trait Client {
+    async fn create_namespace() -> Result<(), Box<dyn Error>>;
+    async fn static_setup(settings: Settings) -> Result<(), Box<dyn Error>>;
+    async fn destroy_namespace() -> Result<(), Box<dyn Error>>;
+    async fn get_node_logs(
+        pod_name: String,
+        since: Option<u32>,
+        with_timestamp: Option<bool>,
+    ) -> Result<String, Box<dyn Error>>;
+    async fn dump_logs(path: String, pod_name: String) -> Result<(), Box<dyn Error>>;
+    async fn upsert_cron_job(minutes: u32) -> Result<(), Box<dyn Error>>;
+    async fn start_port_forwarding(
+        port: u16,
+        identifier: String,
+        namespace: Option<String>,
+    ) -> Result<u16, Box<dyn Error>>;
+    async fn run_command(
+        args: Vec<String>,
+        opts: RunCommandOptions,
+    ) -> Result<RunCommandResponse, Box<dyn Error>>;
+    async fn run_script(
+        identifier: String,
+        script_path: String,
+        args: Vec<String>,
+    ) -> Result<RunCommandResponse, Box<dyn Error>>;
+    async fn spawn_from_def(
+        pod_def: PodDef,
+        files_to_copy: Option<Vec<FileMap>>,
+        keystore: Option<String>,
+        chain_spec_id: Option<String>,
+        db_snapshot: Option<String>,
+    ) -> Result<(), Box<dyn Error>>;
+    async fn copy_file_from_pod(
+        identifier: String,
+        pod_file_path: String,
+        local_file_path: String,
+        container: Option<String>,
+    ) -> Result<(), Box<dyn Error>>;
+    async fn put_local_magic_file(
+        name: String,
+        container: Option<String>,
+    ) -> Result<(), Box<dyn Error>>;
+    async fn create_resource(
+        resourse_def: NameSpaceDef,
+        scoped: bool,
+        wait_ready: bool,
+    ) -> Result<(), Box<dyn Error>>;
+    async fn create_pod_monitor(file_name: String, chain: String) -> Result<(), Box<dyn Error>>;
+    async fn setup_cleaner() -> Result<(), Box<dyn Error>>;
+    async fn is_pod_monitor_available() -> Result<(), bool>;
+    fn get_pause_args(name: String) -> Vec<String>;
+    fn get_resume_args(name: String) -> Vec<String>;
+    async fn restart_node(name: String, timeout: u32) -> Result<(), bool>;
+    async fn get_node_info(
+        identifier: String,
+        port: Option<u16>,
+    ) -> Result<Vec<(String, u32)>, Box<dyn Error>>;
+    async fn get_node_ip(identifier: String) -> Result<String, Box<dyn Error>>;
+    async fn spawn_intro_spector(ws_uri: String) -> Result<(), Box<dyn Error>>;
+    async fn validate_access() -> Result<(), bool>;
+    fn get_logs_command(name: String) -> String;
+}
```

Comment:
I think we should name the main trait Provider because the create is a provider and we have different types of provider PodmanProvider, KubernetesProvider, NativeProvider...

Reply:
I think we also have the opportunity here to refactor and have a more clear boundaries between the `provider` and the `networkNode` (since the responsibility of those are a little mix now).

Reply:
Agreed @l0r1s ;
@pepoviola can you elaborate a bit more on this?

Reply:
Yes, this was more in the spirit of trigger the conversation on how we want to design the interaction between those two `mods`, no need to be addressed/resolved as part of this pr.
In the `ts` version the `client` is used as singleton from the `networkNode` and I think we can improve that design in the new version to have a more clear/loose couple approach.